### PR TITLE
Prevent harvesters getting stuck in gridlock by allowing units to give way.

### LINKED ISSUE #3089: Harvesters can get stuck on unloading
They seem to get stuck at the refinery when multiple harvesters are about to unload at the same refinery. The harvester that should unload does not unload and only unloads after you explicitly order it to. It happens pretty often during games where you have multiple harvesters attached to one refinery.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/244539-harvesters-can-get-stuck-on-unloading?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #3671: Move orders can give away hidden structures
Issuing a move order to a location reveal the presence of an invisible frozen building.

The move target line should indicate the original ordered position, and the actor move as close as they can.

The fix for this probably wants to be tied in to a more general fix for move orders into unpathable terrain, which currently causes the unit to do nothing.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/776457-move-orders-can-give-away-hidden-structures?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10565: Stuck AI harvesters can be a problem
This happens in TS/RA2 where the AI stacks building too close together or builds a refinery which exits to a cliff as the bibs are solely decoration. Harvesters failing to path constantly is expensive.

image
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30137318-stuck-ai-harvesters-can-be-a-problem?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #12576: D2K - AI Harvesters often get stuck
It happens so often that makes games usually uplayable as all Harvesters may get stuck and AI can get no credits.

They stay like that:
harvesters stuck

### LINKED ISSUE #12817: Allied harvesters get stuck when trying to harvest the same tile
I was playing a game last night with my brother, and we teamed up against computers. 

The maps we played were:
1. Morbid Aimless Poseidon
2. Escalations

To reproduce you basically have you and your allies harvesters gather tiberium in the same location. Eventually they will compete for the same tiberium, resulting in harvesters being stuck and never continuing their harvesting, which will cause you to never increase your cash amount, if you are not paying attention to your harvesters.

Maybe a potential fix would be to add some smarts to the harvester to recognize that something is in the way of its path, sort of like the minigunners that can avoid the tiberium, so that the harvester will move to a different tiberium tile instead of being stuck.

### LINKED ISSUE #13260: Harvesters sometimes mine resources it can't harvest
https://www.youtube.com/watch?v=Zl7j4hSTW6g

Steps to reproduce
* Create a map like this to confine ore growth with gems.
* Disallow harvesters to mine gems using map rules
* Build enough harvesters so that they will fight over resources
* Make ore grow really fast.
Harvesters will fight over the narrow ore patch and eventually, they will mine the gem patch.

The sample map is here:
https://github.com/forcecore/OpenRA/tree/HarvDeadlock/mods/ludicrous

### LINKED ISSUE #14752: Harvesters still getting stuck in bleed
Been noticing this while testing my own changes I'm working on and the code for unsticking them doesn't seem to be doing it's job for reasons I don't understand.  Sometimes I'll see several harvesters all parked around an Ore/Gem Mine in the case of RA and also in a few rare cases around refineries.  Doesn't happen too often but when it does the things refuse to move to any other tile than the one they last tried to move to.  Not sure if it's the AI causing this or the Harvesters themselves but I'll see what I can dig up.  That said it happens most often with the AI. Maybe add some kind of failsafe to reroute them to other refineries and resource patches in an ever increasing radius until it finds a new spot it won't get stuck?

Edit:
Images:
https://i.imgur.com/820mfj2.png
https://i.imgur.com/Z785RNp.png
https://i.imgur.com/I7IrkxT.png
https://i.imgur.com/7FZftZC.png
https://i.imgur.com/hxyBwUG.png
https://i.imgur.com/FjN86OR.png

Basically we've got mostly harvesters bugging out either around the resource spawners or around the refineries.  There's also a couple of cases of the AI having issues with land and naval units when using both at the same time but that's a matter for another post.

Here's the test map I was using and the replay: https://drive.google.com/open?id=10yqdS0JgRYnIdOx8GLWkF9afithvfS8o  ,  https://drive.google.com/open?id=1sg8satxJgIMogHhV55xunoSTtvjfN8fO

### LINKED ISSUE #16412: Harvesters refusing all mining orders on bleed
When an ore patch is mined out some harvesters seem to enter a state in which they start to refuse ALL mining orders. One can still issue move orders and set them to a return trip to refineries

Harvester bug

A replay where it happens (last bleed commit #16241):
[OpenRA-2019-04-13T211331Z.zip](https://github.com/OpenRA/OpenRA/files/3076756/OpenRA-2019-04-13T211331Z.zip)
 it was first found in RV.

Also I heavily dislike that when an ore patch is mined out Harvesters start a suicide mission to mine ore all the way across the map

@tovl

### PR DESCRIPTION
**Pull request by tovl**
This fixes the issues with harvesters getting stuck around refineries and resource spawns as well as the majority of other gridlocks caused by units moving in opposite directions where no alternative paths are available.

Fixes #14752
Fixes #12817
Fixes #12576
Fixes #10565
Fixes #3089
Fixes #13260
Fixes #16412 
Fixes #3671

- ~The first commit in this PR contains a general refactor of `HarvestResources` such as was discussed in #16193.~
- The second commit contains changes to the pathfinding logic to allow moving units to give way when their only path is blocked by other moving units, essentially nudging themselves. To facilitate this I also refactored the transient actor checks in the pathfinding code.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/HarvestResource.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/HarvestResource.cs]: Can you please add a comment above this line explaining why we are ignoring the return value?

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/Move.cs]: Style nit: omit the `()` when there's only one argument.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/Move.cs]: Likewise here.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/Move.cs]: And here

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/Move.cs]: And here

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: And here

File: OpenRA.Mods.Common/Pathfinder/PathFinderUnitPathCacheDecorator.cs
Comment (pchote) [inline OpenRA.Mods.Common/Pathfinder/PathFinderUnitPathCacheDecorator.cs]: Do we have any idea about how useful this cache actually is? Ping @teinarss.

* A lot has changed since this was added.  If it doesn't do much in practice anymore then we should remove it completely to help simplify the pathfinder mess.
* If we have lots of repeated queries in the same tick then we should cache everything, but expire the non-`BlockedByActor.None` paths at the end of the tick.

File: OpenRA.Mods.Common/Traits/Mobile.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: I *really* don't like the idea of specific `Move` state leaking into `Mobile`. Can we please find a way to keep this inside `Move`?

File: OpenRA.Mods.Common/Traits/Mobile.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: It may be possible to infer this without using `Path` by looking at `fromCell` and `toCell` or by setting a flag from `EnteringCell` / `FinishedMoving`.

**Review by pchote — COMMENTED** [review]
Some initial thoughts:

This makes a fundamental change to our hottest code paths, so we really need to do some performance comparisons to make sure we don't unreasonably increase the overheads.

The second commit is a really big deal and is going to need some careful testing and analysis, so it might be a good idea to split the first commit into its own PR that we can merge quickly to resolve #16412.

File: OpenRA.Mods.Common/Activities/HarvestResource.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/HarvestResource.cs]: Done.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Mobile.cs
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Mobile.cs]: Okay. Used flags instead.

**Comment by teinarss** [comment]
Looks like the harvesters starting to do a small dance when waiting for docking to the refinery.

nudge

**Comment by pchote** [comment]
I'm excited by the prospect of this superseding `BuildingInfluence` queries in the pathfinding code - i'd expect that buildings should be counted as Immobile and therefore block in cases where we want to allow transient actors.

But on closer inspection it seems that the different `BlockedByActor` cases aren't handled in places like `CrateInfo.GetAvailableSubCell`, so we aren't quite there yet.

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: This is an extremely hot method, so these extra checks could potentially cause a significant perf regression during blob-on-blob battles - need to profile this!

**Review by PunkPun — COMMENTED** [review]
I tried my hardest to get harvesters stuck and failed. GJ

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs
Comment (teinarss) [reply OpenRA.Mods.Common/Traits/World/Locomotor.cs]: Could do some benchmarks around this, but would be nice to have #16328 first

**Comment by pchote** [comment]
#16328 should hopefully be merged this weekend, so it would be good to get this one rebased and for profiling.

**Comment by pchote** [comment]
Needs a rebase, which will also pull in the benchmarking changes to help with profiling.

**Comment by tovl** [comment]
Rebased.

**Comment by teinarss** [comment]
Did a benchmark run and here are the resuts
image

**Comment by tovl** [comment]
What is on the axes?

**Comment by teinarss** [comment]
X axis is ms and y axis is frequency.

**Review by abcdefg30 — COMMENTED** [review]
The benchmark doesn't show hard perf regressions, so this should be good to go? (Regarding the unresolved review comments above.)

**Comment by teinarss** [comment]
> 
> 
> Looks like the harvesters starting to do a small dance when waiting for docking to the refinery.
> 
> nudge

Is there a simple way to minimize this? It would affect the eco a little bit.

**Comment by pchote** [comment]
> The benchmark doesn't show hard perf regressions...

The plot shows that the perf regression is as bad as I feared above, and will be crippling in the large mp games that are currently borderline. We need to implement a cache for the collision checks before we can’t proceed here.

**Comment by abcdefg30** [comment]
Maybe I did misread the graph, I thought red was before this PR. I didn't test it ingame, but the other thing looks to avoid large spikes.

**Comment by teinarss** [comment]
Red is before the PR, but blue is shifted around 5ms (to the right) on the
x axis and more spread out.

It's possible that we can find enough optimization through #16698 to fix
this.

**Comment by pchote** [comment]
We need to branch a new release ASAP, so lets defer this to Next + 1.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Done.

File: OpenRA.Mods.Common/Traits/World/PathFinder.cs
Comment (tovl) [reply OpenRA.Mods.Common/Traits/World/PathFinder.cs]: Done.

**Comment by tovl** [comment]
Rebased.

**Comment by tovl** [comment]
> Looks like the harvesters starting to do a small dance when waiting for docking to the refinery.

Fixed.

**Comment by teinarss** [comment]
> 
> 
> > Looks like the harvesters starting to do a small dance when waiting for docking to the refinery.
> 
> Fixed.

Awesome! Good work!

**Comment by teinarss** [comment]
Looks like units doesn't get pushed away when the unit is ordered to move to a cell in vision. It works when the target is in the shroud.
push

**Comment by tovl** [comment]
This has actually nothing to do with visibility. It will happen whenever the distance is 8 cells or less regardless of the unit's vision range.

This is the much maligned `NearEnough` parameter  and it has the same behaviour on bleed. The purpose is to prevent the `Move` activity from running forever if it is impossible for the unit to reach it's destination and awkward jostling for position when ordering a group of units. Replacing that is out-of-scope for this PR, but I do intend on looking into replacing it with something better for Next+1 in a follow up.

**Comment by reaperrr** [comment]
>  I do intend on looking into replacing it with something better for Next+1 in a follow up.

My idea would be to scale it with number of actors within the selection (though we'd have to figure out how to handle actors with `SharesCell`). Something like 1 actor: 0, 2-9 actors: 1, 10-25 actors: 3 and so on. But yeah, out of scope for this PR.

**Comment by tovl** [comment]
Rebased.

**Comment by tovl** [comment]
Rebasing needs to wait until #16861 is merged.

**Comment by tovl** [comment]
Rebased.

**Comment by tovl** [comment]
Rebased again.

**Comment by pchote** [comment]
@teinarss can you retry your perf testcase?

**Comment by teinarss** [comment]
Log=True
image
Log=False
image

File: OpenRA.Mods.Common/Pathfinder/PathFinderUnitPathCacheDecorator.cs
Comment (teinarss) [reply OpenRA.Mods.Common/Pathfinder/PathFinderUnitPathCacheDecorator.cs]: Im guessing it still is valid for preventing spam click => lag spike.

**Comment by tovl** [comment]
That looks definitely better than before. However, I'm a bit confused: It looks like the total count is lower in the new PR. Shouldn't it be the same? Or am I misinterpreting the graph?

**Comment by pchote** [comment]
The count being lower suggests that units are repathing a lot less, which is a very good thing provided it isn't a mistake.

**Comment by teinarss** [comment]
Can do another run tomorrow if needed.

**Review by teinarss — DISMISSED** [review]
Great work!

**Comment by tovl** [comment]
Rebased.

**Review by RoosterDragon — CHANGES_REQUESTED** [review]
Looks effective and well put together - some minor asks for you.

File: OpenRA.Game/Traits/TraitsInterfaces.cs
Comment (tovl) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Done.

As an aside, I was under the impression that this is actually the most common usage of enums. It is literally the first example in https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/enum . The way it is used throughout the OpenRA code-base, as a way to fit multiple booleans in a byte, seems more unusual to me. 

It is also not always justified. just take the `DamageState` enum just above this as an example. Those categories are clearly mutually exclusive, so there is no reason for them to occupy orthogonal bits. They could just be 0 to 5.

File: OpenRA.Mods.Common/Traits/Mobile.cs
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Mobile.cs]: Done for `IsBlocking`.

`TurnToMove` is referenced by `Mobile.IsLeaving()` so making it a part of `Move` would mean that `Mobile` would need to know about the current activity state. It seems preferable to have activities reference the internals of traits than the other way around.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Move/Move.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Game/Traits/TraitsInterfaces.cs
Comment (RoosterDragon) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: By ordering, I meant the fact your are comparing enum values with `<` or `>` operators instead of just direct equality. Nothing wrong with it - but not that common. Bitwise packing as usually indicated by the `Flags` attribute is something else.

Usually, if I happen across `enum Foo { A, B, C }` I don't really expect consuming code to break if I reorder that to `enum Foo { A, C, B }`. That's what I wanted the comment to guard against (e.g. stop passers by reordering it alphabetically or sticking a new value in "the wrong place", or whatever)

I agree with you r.e. `DamageState` - odd that it's decided for `Flags` there.

File: OpenRA.Mods.Common/Traits/Mobile.cs
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Mobile.cs]: What if `IsLeaving` wasn't defined on `Mobile`? What if it was just a private helper in `Move`? It's the only thing interested in the notion of `Leaving` anyway. Then there's no poking at internals from either side?

Unless there's some semantic reason to keep the "I'm turning as part of a move" flag visible from the trait? I'm only thinking in narrow "Let's not have public get/set properties here since it seems like some private fields would suffice" and maybe there's a wider picture I'm missing.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Mobile.cs]: `IsLeaving` is called by the `Move` activity of the _other_ actor. Making it part of `Move` would mean that an activity on one actor would need to dig in the internals of a different actors activity state. That would be pretty much the worst case scenario in terms of keeping internals internal. `Mobile` is acting like an intermediary here, to provide an interface that other units can access.

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Mobile.cs]: I knew I was missing something - this makes much more sense now, thanks.

**Review by RoosterDragon — APPROVED** [review]

**Comment by RoosterDragon** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/60a89511aae492389559a4135708089cb869d51e)

**Pull request merged**
